### PR TITLE
Render DAGCode in the Grid View as a tab

### LINKED ISSUE #29858: Migrate DAG Code page to Grid Details
- [ ] Use REST API to render DAG Code in the grid view as a tab when a user has no runs/tasks selected
- [ ]  Redirect all urls to new code 
- [ ] delete the old code view

### PR DESCRIPTION
**Pull request by tquazi**
closes: #29858

Using REST API to render DAG Code in the grid view as a tab when a user has no tasks selected.

image

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/www/static/js/dag/details/dagCode/index.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+const DagCode = () => {
+  const dagCodeRef = useRef<HTMLDivElement>(null);
+  const offsetTop = useOffsetTop(dagCodeRef);
+
+  const [lastParsedTime, setLastParsedTime] = useState("");
+  const [dagCode, setDagCode] = useState("");
+
+  const dagApiUrl = getMetaValue("dag_api");
+  if (dagApiUrl) {
+    axios.get<AxiosResponse, API.DAG>(dagApiUrl).then((dagResp) => {
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/dagCode/index.tsx]: Let's move all the fetching logic to react-query. Check out the `js/api` directory for examples.
react query will handle a lot of async logic for us, like loading, errors and caching.

File: airflow/www/static/js/dag/details/dagCode/CodeBlock.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import { PrismLight as SyntaxHighlighter } from "react-syntax-highlighter";
+import { oneLight } from "react-syntax-highlighter/dist/esm/styles/prism";
+import python from "react-syntax-highlighter/dist/esm/languages/prism/python";
+import React, { useRef, useState } from "react";
+import { Box, Button } from "@chakra-ui/react";
+
+SyntaxHighlighter.registerLanguage("python", python);
+interface Props {
+  code: string;
+  language: string;
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/dagCode/CodeBlock.tsx]: Won't language always be python? A constant shouldn't be a prop.

**Review by bbovenzi — CHANGES_REQUESTED** [review]
Looking good!

Few comments:
- Let's move the API logic to react-query and then add some basic handling for loading+error states
- Maybe bump the font-size down? It's quite a bit larger than the rest of the grid view text

File: airflow/www/static/js/dag/details/dagCode/CodeBlock.tsx (hunk 2 above)
Comment (tquazi) [reply airflow/www/static/js/dag/details/dagCode/CodeBlock.tsx]: I have removed the language property

File: airflow/www/static/js/dag/details/dagCode/index.tsx (hunk 1 above)
Comment (tquazi) [reply airflow/www/static/js/dag/details/dagCode/index.tsx]: I have moved the logic to js/api. Code looks clean now :)

**Comment by tquazi** [comment]
> * Maybe bump the font-size down? It's quite a bit larger than the rest of the grid view text

I have reduced the font size to 13.5px. Please have a look and let me know if it looks OK.

File: airflow/www/static/js/dag/details/dagCode/CodeBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  code: string;
+}
+export default function CodeBlock({ code }: Props) {
+  const dagCodeRef = useRef(null);
+  const [codeWrap, setCodeWrap] = useState(false);
+  const toggleCodeWrap = () => setCodeWrap(!codeWrap);
+
+  return (
+    <Box
+      ref={dagCodeRef}
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/dagCode/CodeBlock.tsx]: Are we using this ref for anything?

File: airflow/www/static/js/api/useDagCode.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      const dagSourceApiUrl = getMetaValue("dag_source_api").replace(
+        "_FILE_TOKEN_",
+        fileToken
+      );
+      return axios.get<AxiosResponse, string>(dagSourceApiUrl, {
+        headers: { Accept: "text/plain" },
+      });
+    },
+    {
+      enabled: false,
```
Comment (bbovenzi) [inline airflow/www/static/js/api/useDagCode.ts]:
```suggestion
      enabled: !!dagQuery.data.fileToken
```

And then we can return the entire dagSourceQuery and remove all of the `useEffect`s

File: airflow/www/static/js/dag/details/dagCode/index.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import { useOffsetTop } from "src/utils";
+import Time from "src/components/Time";
+
+import { useDagCode } from "src/api";
+import CodeBlock from "./CodeBlock";
+
+const DagCode = () => {
+  const dagCodeRef = useRef<HTMLDivElement>(null);
+  const offsetTop = useOffsetTop(dagCodeRef);
+  const { lastParsedTime, codeSource = "" } = useDagCode();
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/dagCode/index.tsx]: Let's also pull in `isLoading` and have a spinner.

File: airflow/www/static/js/api/useDagCode.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  });
+
+  const dagQuery = useQuery(
+    ["dagQuery"],
+    () => {
+      const dagApiUrl = getMetaValue("dag_api");
+      return axios.get<AxiosResponse, API.DAG>(dagApiUrl);
+    },
+    {
+      onError: (error: Error) => {
```
Comment (bbovenzi) [inline airflow/www/static/js/api/useDagCode.ts]: We mainly use `errorToast` for edit actions. We don't really need it here. If we want to show an error to a user, we should do it in the dga code react component.

**Review by bbovenzi — CHANGES_REQUESTED** [review]
Looking better! Left a few suggestions.

File: airflow/www/static/js/api/useDagCode.ts (hunk 4 above)
Comment (bbovenzi) [reply airflow/www/static/js/api/useDagCode.ts]: Documentation: https://tanstack.com/query/v3/docs/react/guides/dependent-queries

Comment (tquazi) [reply airflow/www/static/js/api/useDagCode.ts]: I can use the code suggestion to remove the first `useEffects` function. 
I will still need the second one as I am pulling the `last_parsed_date` column from `dagApi` and the `dag_code` from `dagSourceApi` and need to combine them before returning. Please let me know if this needs to change.

File: airflow/www/static/js/dag/details/dagCode/CodeBlock.tsx (hunk 3 above)
Comment (tquazi) [reply airflow/www/static/js/dag/details/dagCode/CodeBlock.tsx]: This is a miss. I have removed it :)

File: airflow/www/static/js/dag/details/dagCode/index.tsx (hunk 5 above)
Comment (tquazi) [reply airflow/www/static/js/dag/details/dagCode/index.tsx]: done

File: airflow/www/static/js/api/useDagCode.ts (hunk 6 above)
Comment (tquazi) [reply airflow/www/static/js/api/useDagCode.ts]: I am now returning the error back to dagCode and handling it there. Let me know if this pattern needs any change.

**Review by bbovenzi — APPROVED** [review]
I added a commit to clean up our API requests. react-query is really cool, it handles most of this logic for us so we don't need to manage loading, errors, or any state and effects ourselves.
A general rule, if you're updating state inside of `useEffect` you're probably doing something wrong. I catch myself doing that too.

**Comment by hduong-mwam** [comment]
Hi, when the user modify the DAG code, is there a way refetch of the new updated code without performing a page refresh?

The question arise when I'm working on implementing the XCom tab in grid view ([issue](https://github.com/apache/airflow/issues/26905)) - which suffers from the same behaviour of not refetching when say a new run occurs and XCom changes.

Thoughts on either a button to refetch or refetch on mount?

**Comment by bbovenzi** [comment]
I would prefer to keep it to a page refetch since changing the DAG code because we're reparsing the dag and those changes could affect any number of ways we render the entire DAG page.

In your PR, isn't an xcom connected to a specific dag run so selecting a new run would then cause another request to the API?

**Pull request merged**